polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp

```diff
@@ -107,6 +109,179 @@ struct RaiseKernelName : public OpRewritePattern<LLVM::AddressOfOp> {
                                        : std::nullopt;
   }
 };
+
+template <typename Derived, typename ConstructorOp, typename TypeTag,
+          bool PostProcess = false>
+class RaiseConstructorBasePattern : public OpRewritePattern<ConstructorOp> {
+public:
+  using OpRewritePattern<ConstructorOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(ConstructorOp constructor,
+                                PatternRewriter &rewriter) const final {
+
+    if (constructor.getArgOperands().empty())
+      return failure();
+
+    // 'this*' is the first argument to the constructor call, if it is a
+    // constructor.
+    auto alloc = dyn_cast_or_null<LLVM::AllocaOp>(
+        constructor.getArgOperands().front().getDefiningOp());
+
+    if (!alloc)
+      return failure();
+
+    assert(alloc.getElemType().has_value() &&
+           "Expecting element type attribute for opaque alloca");
+
+    // Check whether the type allocated for the first argument ('this') matches
+    // the expected type.
+    auto allocTy = *alloc.getElemType();
+    auto structAllocTy = dyn_cast<LLVM::LLVMStructType>(allocTy);
+    if (!structAllocTy || structAllocTy.getName() != TypeTag::getTypeName())
+      return failure();
+
+    if (constructor.getNumResults())
+      // Constructor should not return anything.
+      return failure();
+
+    if (!isConstructor(constructor))
+      // Invoke is not a constructor call.
+      return failure();
+
+    rewriter.create<sycl::SYCLHostConstructorOp>(
+        constructor->getLoc(), constructor.getArgOperands().front(),
+        constructor.getArgOperands().drop_front(1),
+        TypeAttr::get(TypeTag::getTypeFromConstructor(constructor)));
+
+    if constexpr (PostProcess)
+      static_cast<const Derived *>(this)->postprocess(constructor, rewriter);
+
+    rewriter.eraseOp(constructor);
+    return success();
+  }
+
+private:
+  bool isConstructor(ConstructorOp call) const {
+    CallInterfaceCallable callableOp = call.getCallableForCallee();
+    StringRef funcName = callableOp.get<SymbolRefAttr>().getLeafReference();
+
+    llvm::ItaniumPartialDemangler Demangler;
+    Demangler.partialDemangle(funcName.data());
+    if (!Demangler.isCtorOrDtor())
+      return false;
+
+    char *demangled = Demangler.finishDemangle(nullptr, 0);
+    if (!demangled)
+      // Demangling failed
+      return false;
+
+    llvm::StringRef demangledName{demangled};
+    bool isDestructor = demangledName.contains('~');
+    free(demangled);
```

Comment:
I like [my RAII pattern better](https://github.com/intel/llvm/blob/73aaaa349a18293bed361d9ffaf17e9ca0d5c8f4/polygeist/lib/Dialect/Polygeist/Transforms/SYCLRaiseHost.cpp#L62). You can reuse if you like

Reply:
Yes, we can use that. However, I did not want to introduce unnecessary dependencies between the PRs. So, I would suggest, that depending on which PR is merged first, we update the other PR to do so.